File: ngraph/core/src/op/einsum.cpp
// Copyright (C) 2018-2021 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
//

#include <algorithm>
#include <cctype>
#include <ngraph/validation_util.hpp>
#include <string>
#include <unordered_map>

#include "itt.hpp"
#include "ngraph/op/einsum.hpp"

using namespace std;
using namespace ngraph;

NGRAPH_RTTI_DEFINITION(op::v7::Einsum, "Einsum", 7);

op::v7::Einsum::Einsum(const OutputVector& inputs, const std::string& equation)
    : Op(inputs)
    , m_equation(equation)
{
    // normalize input equation by removing extra white-spaces from the equation
    m_equation.erase(std::remove_if(m_equation.begin(), m_equation.end(), ::isspace),
                     m_equation.end());

    constructor_validate_and_infer_types();
}

/// \brief      Check that a subscript contains only alphabetic letters or
/// alphabetic letters with one ellipsis
///
/// \param      subscripts          A subscript to check its format
///
/// \param      is_ellipsis_met     Marker if ellipsis is met in the subscript
///
/// \return     true - correct subscript, false - otherwise
///
bool is_subscript_correct(const std::string& subscript, bool& is_ellipsis_met)
{
    is_ellipsis_met = false;
    auto subscript_length = subscript.length();
    for (size_t ch_idx = 0; ch_idx < subscript_length; ++ch_idx)
    {
        if (is_ellipsis_met == false && ((subscript_length - ch_idx) > 2) &&
            (subscript.substr(ch_idx, 3).compare("...") == 0))
        {
            // mark that ellipsis is met once
            is_ellipsis_met = true;

            // make additional increment since ellipsis consists of three dots.
            ch_idx += 2;
        }
        else if (std::isalpha(subscript[ch_idx]) == 0)
        {
            return false;
        }
    }

    return true;
}

void op::v7::Einsum::parse_equation(const std::string& equation,
                                    std::vector<std::string>& input_subscripts,
                                    std::string& output_subscript)
{
    NGRAPH_OP_SCOPE(v7_Einsum_parse_equation);

    // split equation to input subscripts and an output subscript
    auto pos_output_delimeter = equation.find("->");
    auto input_subscripts_str = equation.substr(0, pos_output_delimeter);

    // split the input subscripts into a vector of input subscripts
    bool is_ellipsis_met = false;
    input_subscripts.clear();
    std::istringstream input;
    input.str(input_subscripts_str);
    for (std::string input_subscript; std::getline(input, input_subscript, ',');)
    {
        bool local_is_ellipsis_met = false;
        // check that input subscript contains only alphabetic letter or ellipsis
        NGRAPH_CHECK(is_subscript_correct(input_subscript, local_is_ellipsis_met),
                     "Input subscript of Einsum equation must consist of either only "
                     "alphabetic letters or alphabetic letters with one ellipsis.");

        // mark that ellipsis is met at least in one input subscript
        if (local_is_ellipsis_met)
        {
            is_ellipsis_met = true;
        }
        input_subscripts.push_back(input_subscript);
    }

    if (pos_output_delimeter == std::string::npos)
    {
        // recover output subscript
        output_subscript = "";
        for (auto const& input_subscript : input_subscripts)
        {
            for (auto const& label : input_subscript)
            {
                if (std::isalpha(label) && output_subscript.find(label) == std::string::npos)
                {
                    output_subscript += label;
                }
            }
        }
        std::sort(output_subscript.begin(), output_subscript.end());
        if (is_ellipsis_met)
        {
            output_subscript = "..." + output_subscript;
        }
    }
    else
    {
        output_subscript = equation.substr(pos_output_delimeter + 2);
        bool output_is_ellipsis_met = false;

        // check that the output subscript has the correct format
        NGRAPH_CHECK(is_subscript_correct(output_subscript, output_is_ellipsis_met),
                     "Output subscript of Einsum equation must consist of either only "
                     "alphabetic letters or alphabetic letters with one ellipsis.");

        // if the ellipsis is met in input subscripts, one ellipsis must be in the output subscript
        NGRAPH_CHECK(is_ellipsis_met == output_is_ellipsis_met,
                     "Output subscript of Einsum equation must contain one ellipsis if "
                     "ellipsis is met in any input subscript.");
    }
}

std::vector<std::string> op::v7::Einsum::extract_labels(const std::string& subscript)
{
    NGRAPH_OP_SCOPE(v7_Einsum_extract_labels);

    std::vector<std::string> labels;
    labels.clear();
    auto subscript_length = subscript.length();
    for (size_t ch_idx = 0; ch_idx < subscript_length; ++ch_idx)
    {
        if (std::isalpha(subscript[ch_idx]))
        {
            labels.push_back(subscript.substr(ch_idx, 1));
        }
        else if (((subscript_length - ch_idx) > 2) &&
                 (subscript.substr(ch_idx, 3).compare("...") == 0))
        {
            labels.push_back("...");
            // make additional increment since ellipsis consists of three dots.
            ch_idx += 2;
        }
        else
        {
            NGRAPH_CHECK(false, "Einsum equation has invalid label.");
        }
    }

    return labels;
}

void op::v7::Einsum::validate_and_infer_types()
{
    NGRAPH_OP_SCOPE(v7_Einsum_validate_and_infer_types);

    // check that Einsum operation has at least one input
    auto num_inputs = get_input_size();
    NODE_VALIDATION_CHECK(this, num_inputs > 0, "Einsum must have at least one input.");

    // check that all inputs have the same type and the type is numeric
    const auto& input_type_0 = get_input_element_type(0);
    NODE_VALIDATION_CHECK(this,
                          input_type_0.is_real() || input_type_0.is_integral_number(),
                          "The input type for Einsum operation must be numeric.");
    for (size_t input_idx = 1; input_idx < num_inputs; ++input_idx)
    {
        const auto& input_type_i = get_input_element_type(input_idx);
        NODE_VALIDATION_CHECK(this,
                              input_type_0 == input_type_i,
                              "Inputs to Einsum operation must have the same type.");
    }

    // check that equation has correct format and extract input and output subscripts
    std::vector<std::string> input_subscripts;
    std::string output_subscript;
    parse_equation(m_equation, input_subscripts, output_subscript);

    // a number of input subscripts must match with a number of input tensors
    NODE_VALIDATION_CHECK(
        this,
        input_subscripts.size() == num_inputs,
        "Equation must contain a number of subscripts equal to a number of Einsum inputs.");

    // create a dictionary with dimension sizes (or ranges in case dynamic shapes) for each label
    // and check their compatibility in case repeating labels
    unordered_map<string, PartialShape> label_to_shape;
    label_to_shape.clear();

    for (size_t input_idx = 0; input_idx < num_inputs; ++input_idx)
    {
        const auto& pshape = get_input_partial_shape(input_idx);
        std::vector<std::string> labels;
        labels = extract_labels(input_subscripts[input_idx]);

        if (pshape.rank().is_static())
        {
            size_t input_rank = pshape.rank().get_length();
            // check that a rank is greater or equal to a number of labels
            // these numbers are always equal if there is no ellipsis in the subscript
            NODE_VALIDATION_CHECK(
                this,
                input_rank >= labels.size(),
                "Input rank must be greater or equal to a number of labels in the "
                "corresponding input subscript.");

            for (size_t label_ind = 0, dim_ind = 0;
                 label_ind < labels.size() && dim_ind < input_rank;
                 ++label_ind)
            {
                auto const& label = labels[label_ind];
                if (label.compare("...") == 0)
                {
                    size_t num_broadcasted_dims = input_rank - labels.size() + 1;
                    auto current_sub_pshape = PartialShape(std::vector<Dimension>(
                        pshape.begin() + dim_ind, pshape.begin() + dim_ind + num_broadcasted_dims));
                    if (label_to_shape.find(label) == label_to_shape.end())
                    {
                        label_to_shape[label] = current_sub_pshape;
                    }
                    else
                    {
                        bool is_broadcast_success =
                            PartialShape::broadcast_merge_into(label_to_shape[label],
                                                               current_sub_pshape,
                                                               op::AutoBroadcastType::NUMPY);
                        NODE_VALIDATION_CHECK(this,
                                              is_broadcast_success,
                                              "Input dimensions labeled with ellipsis for Einsum "
                                              "must be broadcastable.");
                    }
                    dim_ind += num_broadcasted_dims;
                }
                else
                {
                    if (label_to_shape.find(label) == label_to_shape.end())
                    {
                        label_to_shape[label] = PartialShape{pshape[dim_ind]};
                    }
                    else
                    {
                        NODE_VALIDATION_CHECK(
                            this,
                            label_to_shape[label].compatible(PartialShape{pshape[label_ind]}),
                            "Different input dimensions indicated by the same labels for Einsum "
                            "must be compatible.");
                        PartialShape::merge_into(label_to_shape[label],
                                                 PartialShape{pshape[dim_ind]});
                    }
                    ++dim_ind;
                }
            }
        }
        else
        {
            for (auto const& label : labels)
            {
                NODE_VALIDATION_CHECK(this,
                                      label != "...",
                                      "The subscript corresponding to a dynamic rank input must "
                                      "not contain ellipsis.");

                if (label_to_shape.find(label) == label_to_shape.end())
                {
                    label_to_shape[label] = PartialShape{Dimension::dynamic()};
                }
            }
        }
    }

    // compute the output shape
    std::vector<std::string> output_labels;
    output_labels = extract_labels(output_subscript);
    std::vector<Dimension> output_pshape_vector;

    for (auto const& output_label : output_labels)
    {
        NODE_VALIDATION_CHECK(this,
                              label_to_shape.find(output_label) != label_to_shape.end(),
                              "Label in output subscript of Einsum equation must enter at least "
                              "one input subscript.");
        output_pshape_vector.insert(output_pshape_vector.end(),
                                    label_to_shape[output_label].begin(),
                                    label_to_shape[output_label].end());
    }
    set_output_type(0, input_type_0, PartialShape(output_pshape_vector));
}

bool op::v7::Einsum::visit_attributes(AttributeVisitor& visitor)
{
    NGRAPH_OP_SCOPE(v7_Einsum_visit_attributes);
    visitor.on_attribute("equation", m_equation);
    return true;
}

shared_ptr<Node> op::v7::Einsum::clone_with_new_inputs(const OutputVector& new_args) const
{
    NGRAPH_OP_SCOPE(v7_Einsum_clone_with_new_inputs);
    check_new_args_count(this, new_args);
    return make_shared<v7::Einsum>(new_args, m_equation);
}

File: tools/mo/openvino/tools/mo/ops/einsum.py
# Copyright (C) 2018-2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import re

import numpy as np

from openvino.tools.mo.front.common.partial_infer.utils import shape_array
from openvino.tools.mo.graph.graph import Node, Graph
from openvino.tools.mo.ops.op import Op
from openvino.tools.mo.utils.broadcasting import bi_directional_shape_broadcasting


class Einsum(Op):
    op = 'Einsum'
    enabled = False

    def __init__(self, graph: Graph, attrs: dict):
        mandatory_props = {
            'type': self.op,
            'op': self.op,
            'version': 'opset7',
            'infer': self.infer,
            'out_ports_count': 1,
        }
        super().__init__(graph, mandatory_props, attrs)

    def backend_attrs(self):
        return ['equation']

    @staticmethod
    def is_label_elsewhere(input_subscripts: list, label_to_check: str, excluded_subscript_inds: list) -> bool:
        """
        Check if the given label is met in input subscripts excluding ones specified by a list of indices
        excluded_subscript_inds

        :param input_subscripts: input subscripts among which to check if the label is met
        :param label_to_check: a label to check
        :param excluded_subscript_inds: indices of input subscripts to be excluded for this check
        :return: True - met, False - otherwise
        """
        for ind, input_subscript in enumerate(input_subscripts):
            if ind not in excluded_subscript_inds and label_to_check in input_subscript:
                return True
        return False

    @staticmethod
    def parse_equation(node_name: str, equation: str) -> (list, str):
        """
        Parse Einsum equation and check that its format is correct to make sure that
        all input subscripts consists of only alphabetic letters or alphabetic letters with one ellipsis.
        In case of implicit mode the method recovers the right-hand part.

        :param node_name: Einsum node name for which to parse an equation
        :param equation: Equation to be parsed and checked
        :return: A tuple of a list of input subscripts and output subscript
        """
        # normalize equation by removing white-spaces
        equation = equation.strip()

        # split equation into the left and right hands
        splitted_equation = equation.split('->')
        assert len(splitted_equation) <= 2, "Einsum node {} has `equation` of incorrect format".format(node_name)

        # split left-hand side of the equation and check a format of input subscripts
        input_subscripts = splitted_equation[0]
        input_subscripts_list = input_subscripts.split(',')

        # prepare pattern to check a format of subscripts
        subscript_pattern = re.compile("^[a-zA-Z]*(\\.\\.\\.){0,1}[a-zA-Z]*$")
        ellipsis_pattern = re.compile("\\.\\.\\.")

        is_ellipsis_met = False
        for input_subscript in input_subscripts_list:
            assert re.match(subscript_pattern, input_subscript) is not None, \
                "Einsum node {} has `equation` with incorrect input subscript: {}".format(node_name, input_subscript)
            is_ellipsis_met = is_ellipsis_met or re.search(ellipsis_pattern, input_subscript)

        if len(splitted_equation) == 2:
            output_subscript = splitted_equation[1]
            assert re.match(subscript_pattern, output_subscript), \
                "Einsum node {} has `equation` with incorrect output subscript: {}".format(node_name, output_subscript)
            # if ellipsis is met, the output subscript must contain it as well
            if is_ellipsis_met:
                assert re.search(ellipsis_pattern, output_subscript), \
                    "The output subscript of Einsum node {} must contain ellipsis".format(node_name)
        elif len(splitted_equation) == 1:
            # recover output subscript in case implicit mode
            output_subscript = ""
            for ind, input_subscript in enumerate(input_subscripts_list):
                labels = Einsum.extract_subscript_labels(node_name, input_subscript)
                for label in labels:
                    if Einsum.is_label_elsewhere(input_subscripts_list, label, [ind]) is False:
                        output_subscript += label
            output_subscript = ''.join(sorted(list(set(output_subscript) - {'.'})))
            if is_ellipsis_met:
                output_subscript = "..." + output_subscript
        else:
            assert False, "Einsum node {} equation has incorrect format. " \
                          "It must be in either explicit or implicit mode.".format(node_name)

        return input_subscripts_list, output_subscript

    @staticmethod
    def normalize_equation(node_name: str, equation: str) -> str:
        """
        Recover explicit mode of equation.

        :param node_name: Einsum node name for which to recover explicit mode
        :param equation: Einsum equation to recover explicit mode
        :return: Recovered equation in explicit mode
        """
        input_subscripts_list, output_subscript = Einsum.parse_equation(node_name, equation)
        return ','.join(input_subscripts_list) + "->" + output_subscript

    @staticmethod
    def extract_subscript_labels(node_name: str, subscript: str) -> list:
        """
        Extract labels for given subscript. Each label can be either alphabetic letter or ellipsis

        :param node_name: Einsum node name
        :param subscript: Given subscript
        :return: A list of labels
        """
        labels = []
        len_subscript = len(subscript)
        label_ind = 0
        while label_ind < len_subscript:
            if subscript[label_ind].isalpha():
                labels.append(subscript[label_ind])
                label_ind += 1
            elif len_subscript - label_ind > 2 and subscript[label_ind:label_ind + 3] == "...":
                labels.append("...")
                label_ind += 3
            else:
                assert False, "Einsum node {} has `equation` with incorrect subscript: {}".format(node_name, subscript)
        return labels

    @staticmethod
    def adjust_equation_with_NCHW_layout(node_name: str, equation: str, input_ranks: list, output_rank: int,
                                         input_correct_layout_mask: list, output_correct_layout_mask: bool) -> (
            str, list, bool):
        """
        In order to satisfy NCHW layout, subscripts for tensors with rank greater than three must be adjusted by moving labels
        of the last dimension to the second position in the subscript. There is an exception for such tensors when
        the label is ellipsis and it covers multiple tail dimensions. The method returns equation with adjusted subscripts
        to NCHW layout along with a boolean mask to indicate which subscripts are adjusted.

        :param node_name: Einsum node name for which equation is adjusted
        :param equation: Equation to be adjusted
        :param input_ranks: a list of input ranks
        :param output_rank: output rank
        :return: adjusted equation, boolean mask for inputs, and boolean flag if output subscript is adjusted
        """
        is_inputs_adjusted = []
        input_subscripts, output_subscript = Einsum.parse_equation(node_name, equation)
        num_inputs = len(input_ranks)
        assert len(input_subscripts) == num_inputs, "The number of inputs must match a number " \
                                                    "of input subscripts"
        assert len(input_correct_layout_mask) == num_inputs, "The number of inputs must match a number " \
                                                             "elements in input_correct_layout_mask list"

        # permute labels in input subscripts and mark inputs for which inference in NCHW layout is acceptable
        # in case ellipsis covering multiple dimensions in the end, the permutation is impossible
        # so the corresponding input must be in the original format (NHWC)
        permuted_input_subscripts = []
        for input_ind in range(num_inputs):
            input_subscript = input_subscripts[input_ind]
            input_rank = input_ranks[input_ind]
            labels = Einsum.extract_subscript_labels(node_name, input_subscript)
            num_broadcasted_dims = input_rank - len(labels) + 1
            if input_correct_layout_mask[input_ind]:
                is_inputs_adjusted.append(True)
            elif input_rank > 3 and (labels[-1] != "..." or labels[-1] == "..." and num_broadcasted_dims == 1):
                is_inputs_adjusted.append(True)
                labels.insert(1, labels[-1])
                del labels[-1]
            else:
                is_inputs_adjusted.append(False)
            permuted_input_subscript = ''.join(labels)
            permuted_input_subscripts.append(permuted_input_subscript)

        # perform the same procedure for the output subscript as for the inputs subscripts
        labels = Einsum.extract_subscript_labels(node_name, output_subscript)
        num_broadcasted_dims = output_rank - len(labels) + 1
        if output_correct_layout_mask:
            is_output_adjusted = True
        elif output_rank > 3 and (labels[-1] != "..." or labels[-1] == "..." and num_broadcasted_dims == 1):
            is_output_adjusted = True
            labels.insert(1, labels[-1])
            del labels[-1]
        else:
            is_output_adjusted = False
        permuted_output_subscript = ''.join(labels)

        # concatenate the left and right hands of the resulted equation
        left_hand = ','.join(permuted_input_subscripts)
        right_hand = permuted_output_subscript
        permuted_equation = left_hand + "->" + right_hand
        return permuted_equation, is_inputs_adjusted, is_output_adjusted

    @staticmethod
    def infer(node: Node):
        node_name = node.soft_get('name', node.id)
        connected_in_ports = [port for port in node.in_ports().values() if not port.disconnected()]
        num_inputs = len(connected_in_ports)
        assert node.has_valid('equation'), "Einsum node {} must contain `equation` attribute".format(node_name)
        equation = node.equation

        # parse the equation and extract input and output subscripts
        input_subscripts, output_subscript = Einsum.parse_equation(node_name, equation)

        # check that each operand has the corresponding input subscript
        assert len(input_subscripts) == num_inputs, "The number of input operands of Einsum node {} " \
                                                    "must match the number of input subscripts " \
                                                    "in `equation`".format(node_name)

        # check compatibility of dimension sizes with the same label and generate a dictionary of shapes for labels
        label_to_shape = {}
        for input_ind in range(num_inputs):
            input_shape = node.in_port(input_ind).data.get_shape()
            input_subscript = input_subscripts[input_ind]
            labels = Einsum.extract_subscript_labels(node_name, input_subscript)
            num_dims = len(input_shape)
            num_labels = len(labels)
            num_broadcasted_dims = num_dims - num_labels + 1
            dim_ind = 0
            label_ind = 0
            while label_ind < num_labels and dim_ind < num_dims:
                label = labels[label_ind]
                if label == "...":
                    sub_shape = input_shape[dim_ind:dim_ind + num_broadcasted_dims]
                    if label in label_to_shape.keys():
                        common_shape = bi_directional_shape_broadcasting(sub_shape, label_to_shape[label])
                        assert common_shape is not None, "The dimensions labeled of ellipsis must be broadcastable " \
                                                         "for Einsum node {}".format(node_name)
                        label_to_shape[label] = common_shape
                    else:
                        label_to_shape[label] = sub_shape
                    dim_ind += num_broadcasted_dims
                else:
                    dim_size = input_shape[dim_ind]
                    sub_shape = shape_array([dim_size])
                    assert label not in label_to_shape.keys() or np.array_equal(label_to_shape[label], sub_shape), \
                        "Sizes of dimensions with the same label of Einsum node {} " \
                        "must be compatible".format(node_name)
                    label_to_shape[label] = sub_shape
                    dim_ind += 1
                label_ind += 1

        # generate output shape based on the output subscript
        output_shape = shape_array([])
        labels = Einsum.extract_subscript_labels(node_name, output_subscript)
        for label in labels:
            assert label in label_to_shape.keys(), "The label in the output subscript must appear" \
                                                   " in input subscripts in equation {} " \
                                                   "of Einsum node {}".format(equation, node_name)
            output_shape = np.ma.concatenate((output_shape, label_to_shape[label]))

        node.out_port(0).data.set_shape(output_shape)
